bigflow/scaffold/infra.py

```diff
@@ -0,0 +1,159 @@
+        gcp_project_id,
+        bigflow_project_name,
+        environment_name,
+        region)
+
+    @secure_destroy(f'{nat_name} (Google Cloud NAT)')
+    def destroy() -> None:
+        logger.info(run_process([
+            'gcloud', 'compute',
+            '--project', gcp_project_id,
+            'routers', 'nats', 'delete', nat_name,
+            '--router', router_name,
+            '--quiet'
+        ]))
+        router_destroyer()
+
+    def create() -> None:
+        router_creator()
+        logger.info(run_process([
+            'gcloud', 'compute',
+            '--project', gcp_project_id,
+            'routers', 'nats', 'create', nat_name,
+            f'--router={router_name}',
+            '--auto-allocate-nat-external-ips',
+            '--nat-all-subnet-ip-ranges',
+            '--enable-logging'
+        ]))
+
+    return nat_name, destroy, create
+
+
+def _composer_create_command(
+        composer_name: str,
+        gcp_project_id: str,
+        region: str,
+        zone: str,
+        environment_name: str) -> tp.List[str]:
+    return [
+        'gcloud', 'composer',
+        '--project', gcp_project_id,
+        'environments', 'create', composer_name,
+        f'--location={region}',
+        f'--zone={zone}',
+        f'--env-variables=env={environment_name}',
+        f'--machine-type=n1-standard-2',
+        f'--node-count=3',
+        f'--python-version=3',
+        f'--enable-ip-alias',
+        f'--network=default',
+        f'--subnetwork=default',
+        f'--enable-private-environment',
+        f'--enable-private-endpoint',
+    ]
+
+
+def cloud_composer(
+        gcp_project_id: str,
+        bigflow_project_name: str,
+        environment_name: str = 'dev',
+        region: str = 'europe-west1',
+        zone: str = 'europe-west1-d') -> tp.Tuple[str, InfrastructureDestroyer, InfrastructureCreator]:
+    composer_name = f'{BIGFLOW_INFRA_PREFIX}-composer-{bigflow_project_name}-{environment_name}-{region}'
+    cloud_nat_name, cloud_nat_destroyer, cloud_nat_creator = _create_cloud_nat(
+        gcp_project_id, bigflow_project_name, environment_name, region)
+
+    @secure_destroy(f'{composer_name} (Google Cloud Composer)')
+    def destroy() -> None:
+        logger.info(run_process([
+            'gcloud', 'composer',
+            'environments', 'delete', composer_name,
+            '--location', region,
+            '--project', gcp_project_id,
+            '--quiet'
+        ]))
+        cloud_nat_destroyer()
+
+    def create() -> None:
+        cloud_nat_creator()
+        logger.info(run_process(_composer_create_command(
+            composer_name, gcp_project_id, region, zone, environment_name)))
+
+    return composer_name, destroy, create
+
+
+def try_create(
+        name: str,
+        destroyer: InfrastructureDestroyer,
+        creator: InfrastructureCreator) -> None:
+    try:
+        creator()
+    except Exception as e:
```

Comment:
```python
except Exception:
        logger.exception(f"Can't create {name}. Trying to destroy leftovers.")
        destroyer()
```

Reply:
I would mix those. Logging real exception message is really helfpul.

Reply:
`logger.exception` prints both log message and exception information (traceback + exception text). [example](https://onecompiler.com/python/3wshm4c2b)
In contrast with Java there is no need to pass exception object into logger explicitly.
But you still could do it by adding named arg 'exc_info` [example](https://onecompiler.com/python/3wshkm2c4)

Reply:
done